client/consensus/manual-seal/src/seal_new_block.rs

```diff
@@ -0,0 +1,173 @@
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Block sealing utilities
+
+use crate::{Error, rpc};
+use std::sync::Arc;
+use sp_runtime::{
+	traits::{Block as BlockT, Header as HeaderT},
+	generic::BlockId,
+};
+use transaction_pool::txpool;
+use rpc::CreatedBlock;
+
+use consensus_common::{
+	self, BlockImport, Environment, Proposer,
+	ForkChoiceStrategy, BlockImportParams, BlockOrigin,
+	ImportResult, SelectChain,
+	import_queue::BoxBlockImport,
+};
+use sp_blockchain::HeaderBackend;
+use client_api::backend::Backend as ClientBackend;
+use hash_db::Hasher;
+use std::collections::HashMap;
+use std::time::Duration;
+use std::marker::PhantomData;
+
+/// params for sealing a new block
+pub struct SealBlockParams<'a, B: BlockT, C, CB, E, P: txpool::ChainApi, H> {
+	/// if true, empty blocks(without extrinsics) will be created.
+	/// otherwise, will return Error::EmptyTransactionPool.
+	pub create_empty: bool,
+	/// instantly finalize this block?
+	pub finalize: bool,
+	/// specify the parent hash of the about-to-created block
+	pub parent_hash: Option<<B as BlockT>::Hash>,
+	/// sender to report errors/success to the rpc.
+	pub sender: rpc::Sender<CreatedBlock<<B as BlockT>::Hash>>,
+	/// transaction pool
+	pub pool: Arc<txpool::Pool<P>>,
+	/// client backend
+	pub back_end: Arc<CB>,
+	/// Environment trait object for creating a proposer
+	pub env: &'a mut E,
+	/// SelectChain object
+	pub select_chain: &'a C,
+	/// block import object
+	pub block_import: &'a mut BoxBlockImport<B>,
+	/// inherent data provide
+	pub inherent_data_provider: &'a inherents::InherentDataProviders,
+	/// phantom data to pin the hasher type.
+	pub _phantom: PhantomData<H>
+}
+
+/// seals a new block with the given params
+pub async fn seal_new_block<B, C, CB, E, P, H>(params: SealBlockParams<'_, B, C, CB, E, P, H>)
+	where
+		B: BlockT,
+		H: Hasher<Out=<B as BlockT>::Hash>,
+		CB: ClientBackend<B, H>,
+		E: Environment<B>,
+		E::Error: std::fmt::Display,
+		<E::Proposer as Proposer<B>>::Error: std::fmt::Display,
+		P: txpool::ChainApi<Block=B, Hash=<B as BlockT>::Hash>,
+		C: SelectChain<B>,
+{
+	let SealBlockParams {
+		create_empty,
+		finalize,
+		pool,
+		parent_hash,
+		sender,
+		back_end,
+		select_chain,
+		block_import,
+		env,
+		inherent_data_provider,
+		..
+	} = params;
+
+	if pool.status().ready == 0 && !create_empty {
+		return rpc::send_result(sender, Err(Error::EmptyTransactionPool));
+	}
+
+	// get the header to build this new block on.
+	// use the parent_hash supplied via `EngineCommand`
+	// or fetch the best_block.
+	let header = match parent_hash {
+		Some(hash) => {
+			back_end.blockchain().header(BlockId::Hash(hash))
+				.map_err(Error::from)
+				.and_then(|header| {
+					match header {
+						Some(header) => Ok(header),
+						None => Err(Error::BlockNotFound)
+					}
+				})
+		}
+		None => select_chain.best_chain().map_err(Error::from)
+	};
+
+	let header = match header {
```

Comment:
The error handling here is tedious, can't we use `map_err` and `?` operator and call` rpc::send_result` in single place?

Reply:
that would be possible if `seal_new_block` returned a `Result`

Reply:
It's possible now too, you just need to introduce an extra function:
```
fn seal_new_block(...) -> () {
  fn helper_function(..) -> Result<(), _> {
     ... // current logic using `.map_err` and `?`
  } 

  if let Err(e) = helper_function(...) {
     rpc::send_result(e);
  }  
}